### PR TITLE
Dynamically Load Star Systems

### PR DESCRIPTION
**Pull request by nixpulvis**
See #20.

### CODE REVIEW

File: starmap/src/ui.rs [hunk 1]
```diff
@@ -1,60 +1,95 @@
...
+        ui.add(egui::Slider::new(&mut radius.0, 0.0..=250.0).text("Radius"));
+
+        ui.checkbox(&mut always_fetch.0, "Always Fetch Systems");
+
+        let last_value = always_despawn.0;
+        ui.checkbox(&mut always_despawn.0, "Always Despawn Systems");
+        if always_despawn.0 && always_despawn.0 != last_value {
+            for entity in systems_query.iter() {
+                commands.entity(entity).despawn_recursive();
+            }
```
Comment (nixpulvis) [inline starmap/src/ui.rs:35]: I think in general, I'd like to have a better way to react to the settings being changed here. I like that I can easily set these booleans here and read them while fetching and spawning systems, but I'm having a little trouble making them update the state of the game immediately.

With this logic, you end up with empty space after enabling this setting. Without this logic, enabling the checkbox doesn't cleanup the old stars.

Thinking out loud, perhaps this is a good case for another event. Instead of doing the cleanup here, we could have a `Cleanup` event which in handled elsewhere and which is used in general as the camera is moved far from things.

Thinking in physical terms, as we prepare for `big_space`. Think of a speed limit many, many times the speed of light. The default `Cleanup` handler should despawn star systems farther than you could reach in the time it would take to retrieve the data... Or something like that.

File: starmap/src/systems.rs [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+                center.z as f64,
+            ];
+            System::fetch_in_range_of_point(&db, radius, cent).await.unwrap_or_default()
+        });
+        loaded_regions.centers.insert(center);
+        tasks.regions.insert(center, task);
+    }
+}
+
+const FACTION_HACK: IVec3 = IVec3::splat(989);
```
Comment (jbrod22) [inline starmap/src/systems.rs:125]: What does this do?

File: starmap/src/systems.rs [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+    if always_fetch.0 {
+        fetch_around_camera(
+            &camera_query,
+            &mut loaded_regions,
+            &mut tasks,
+            &mut radius,
+            &db);
+    }
+
+    for event in search_events.read() {
```
Comment (jbrod22) [inline starmap/src/systems.rs:67]: What are the pros/cons of reading events this way vs. separate systems?

File: starmap/src/systems.rs (hunk 2 above)
Comment (nixpulvis) [reply starmap/src/systems.rs:125]: So let me try to explain this... Basically we need a way to identify fetched systems. We have both `FetchTasks` and `LoadedRegions` which either contain a `HashMap<IVec3, Task<Vec<System>>>` or a `HashSet<IVec3>` respectively. We use the current camera's center as that `IVec3`.

So when we move the camera (for example) we check if the new camera center is in the hash/set before fetching...
https://github.com/nixpulvis/galos/blob/5b63b7ff8aa94629bd4e89729c8323355e667641/starmap/src/systems.rs#L104-L108

OK, now why the hack. Well the loaded regions are identified by the center... this works when you're looking at the bubble "spyglass" view of the systems, i.e. a sphere around the camera. But when a user fetches a faction's systems, or plots a route, now the center of the camera isn't what identifies that fetch. Really what identifies it is the fact it's a faction or route fetch, so that's where I get the desired `enum`, I don't bother trying to hash the faction/route data into the existing `IVec3` because until I make it an `enum` there's always going to be an increased chance of collisions, like there are now.

https://github.com/nixpulvis/galos/blob/5b63b7ff8aa94629bd4e89729c8323355e667641/starmap/src/systems.rs#L133

https://github.com/nixpulvis/galos/blob/5b63b7ff8aa94629bd4e89729c8323355e667641/starmap/src/systems.rs#L154

https://github.com/nixpulvis/galos/blob/5b63b7ff8aa94629bd4e89729c8323355e667641/starmap/src/systems.rs#L221

https://github.com/nixpulvis/galos/blob/5b63b7ff8aa94629bd4e89729c8323355e667641/starmap/src/systems.rs#L228

I don't seem to bother saving the factions and routes in the `loaded_regions` which I think is just as well for now.

TLDR; I needed a way to decide when faction and route searches should do stuff **after** the task is done. The `fetch` system gets `Searched` events which know what kind of search they are, but the `spawn` system just processes tasks spawned by `fetch`.

File: starmap/src/systems.rs (hunk 3 above)
Comment (nixpulvis) [reply starmap/src/systems.rs:67]: Do you mean a separate system from one that does the `if always_fetch.0 { ... }` because that's all this system is doing otherwise. Basically the system logic is, see if we need to fetch around the camera, then see if we have any other events that need fetching for.

I felt like this was a pretty logical single system, especially since the user is unlikely to be triggering `Searched` events while the camera is moving I think. Unless I'm missing something, I don't expect a bottleneck here.

**Review by nixpulvis — COMMENTED** [review]
Leaving way to many comments for myself and answering a couple questions.

File: starmap/src/main.rs [hunk 4]
```diff
@@ -50,17 +59,30 @@ fn main() {
             color: Color::default(),
             brightness: 1000.0,
         })
+        .insert_resource(DatabaseResource(db))
+        .insert_resource(systems::SpyglassRadius(50.))
+        .insert_resource(systems::AlwaysFetch(true))
```
[the remark this answers is not in the document]
Comment (nixpulvis) [reply starmap/src/main.rs:49]: I think it all makes sense for now. We'll see if things change when I refactor the system fetching to be in cubes and another layer of caching is done at the entity level with marker components to control visibility.

File: starmap/src/main.rs [hunk 5]
```diff
@@ -50,17 +59,30 @@ fn main() {
...
-        .add_systems(Update, ui::faction_search)
-        .add_systems(Update, ui::route_search)
+        .add_systems(Update, systems::fetch)
+        .add_systems(Update, systems::spawn)
+
+        .add_systems(Update, ui::settings)
+        .add_systems(Update, ui::search.after(ui::settings))
+        .add_systems(Update, ui::route.after(ui::search))
+
+        .add_systems(Update, search::system)
```
[the remark this answers is not in the document]
Comment (nixpulvis) [reply starmap/src/main.rs:69]: This clearly belongs in `systems::fetch`:

https://github.com/nixpulvis/galos/blob/1360bcd20dcafa5ca9392aa70d28ee1fb1934a5f/starmap/src/systems.rs#L73

[the remark this answers is not in the document]
Comment (nixpulvis) [reply starmap/src/main.rs:69]: The issue with this is that the linked `fetch` system doesn't actually get the ` `System` back, but spawns a task... So where this code really belongs is here:
https://github.com/nixpulvis/galos/blob/5084bc60ff41ba2dc8df4c2a246a8635658bd0d9/starmap/src/systems.rs#L245-L254

Because, like the `route` post-fetch spawn needs to do something, we too need to do something (send a `MoveCamera`) event.

**Comment by nixpulvis** [comment]
Found a bug. After searching for a star system, changing the spyglass radius doesn't trigger a fetch.

**Comment by nixpulvis** [comment]
Hilarious. Because Sol is at (0,0,0) and that's where I was testing this, it's the only place that fails and has nothing to do with the search (good). The reason for this is here:
https://github.com/nixpulvis/galos/blob/6ee365a910f155a245220d210fdae040db652987/starmap/src/systems.rs#L115

When center is (0,0,0) dividing by scale doesn't change the region so nothing new loads. This will be fixed when I expand the definition for the elements of `FetchTasks` and `LoadedRegions`.

File: starmap/src/systems.rs (hunk 2 above)
Comment (nixpulvis) [reply starmap/src/systems.rs]: @jbrod22 In case you care, here's the fix for these hack: 097cb4e9a9ea27bc731f87b56a5060fca8388078

File: starmap/src/main.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (nixpulvis) [reply starmap/src/main.rs:75]: I'm going to deal with this later, it's stubbed out in `FetchIndex` along with the `FetchIndex::View` for eventual full camera view renders!

**Comment by nixpulvis** [comment]
This should be good to go I think. I'll probably do another read over it though before I merge.

File: starmap/src/ui.rs (hunk 1 above)
Comment (nixpulvis) [reply starmap/src/ui.rs:54]: I left some notes.

https://github.com/nixpulvis/galos/blob/097cb4e9a9ea27bc731f87b56a5060fca8388078/starmap/src/systems.rs#L190-L194

https://github.com/nixpulvis/galos/blob/097cb4e9a9ea27bc731f87b56a5060fca8388078/starmap/src/systems.rs#L217-L223

https://github.com/nixpulvis/galos/blob/097cb4e9a9ea27bc731f87b56a5060fca8388078/starmap/src/systems.rs#L262
